### PR TITLE
[ci/docker] Move CI + Docker to Python 3.8 [build_base]

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Implementing #34863, this PR updates our CI and Docker images to use Python 3.8 as the default Python version.

It also updates CUDA version numbers and removes stale CUDA images.

The compatibility test becomes and explicit Python 3.7 compatibility test. Currently the tests we are running are (with some filters):

- python/ray/tests/horovod
- python/ray/tests/lightgbm
- python/ray/tests/xgboost
- python/ray/tests/ray_lightning
- python/ray/tests/ml_py37_compat (keras + torch)
- python/ray/air
- python/ray/train (AIR tests)
- python/ray/data (AIR tests)

We may want to add more tests here from other libraries, e.g. rllib (cc @sven1977 @kouroshHakha) and Serve (cc @iycheng @edoakes)

### Technical details

Due to a bug in libstdc++ (e.g. tracked here: https://github.com/ray-project/ray/issues/33858), the Ray processes sometimes error out. It looks like this occurred rarely in Python 3.7, but after the switch to Python 3.8 we see it more often in some of our tests.

Note that this is an ephemeral failure and not due to specific tests. For instance, we saw `test_trial_runner_3.py` fail regularly, but it was always different tests. It seemed that just by the sheer amount of tests, and the quick succession of starting/stopping Ray clusters, one of these tests was bound to fail.

For those tests, we introduced an intermediate solution using pytest reruns, namely the `--reruns 3` option. This specifies reruns on the subtest level, as opposed to bazels test-suite level. The reruns option was able to make these flaky tests green. However, this undermines our flaky test tracking, as actually flaky tests will be detected less often. After moving to Ubuntu 22 and a new stdlibc++, we should remove these subtest-reruns.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ollie-iterators** [comment]
The link error should be fixed by merging in recent changes from the main branch

**Comment by kouroshHakha** [comment]
@krfricke we should add rllib ci tests too? (At least in the interim state) As much as we would like to bump up the python version some of the RLlib customers are still on 3.7.

**Comment by krfricke** [comment]
@kouroshHakha the main question is which tests you'd like to run on 3.7. Ideally it's covers core functionality to make sure the main workflows continue to work. It's probably feasible to run all existing rllib tests on both 3.7 and 3.8.

File: python/ray/tests/test_cli.py
Comment (krfricke) [inline python/ray/tests/test_cli.py:88]: I've bulk-added these as otherwise some of the tests were failing. Maybe we can clean this up in a follow-up

File: python/requirements/data_processing/requirements.txt
Comment (ArturNiederfahrenhorst) [inline python/requirements/data_processing/requirements.txt:12]: Why do we differentiate between Python versions here if both groups pin at the same version?

File: .buildkite/pipeline.build.yml
Comment (ArturNiederfahrenhorst) [inline .buildkite/pipeline.build.yml:669]: Why do we pick these tests for RLlib? They seem like a random subset of RLlib tests to me.

If we test for RLlib on 3.7, from my point of view, "RLlib: RLModule tests" and "RLlib: Examples" may be included to also test our new RLModule/Learner stack on 3.7 until it is fully deprecated. Examples also cover some extravagant stuff.

For RLlib, I would probably not fix issues that arise from these just for 3.7 if anything comes up, but it helps to be aware to better handle issues.

**Comment by jjyao** [comment]
> However, this undermines our flaky test tracking, as actually flaky tests will be detected less often.

Are you saying go/flaky will be affected by this? like as long as one run succeeds, it's green on go/flaky?

File: .buildkite/pipeline.build.yml
Comment (krfricke) [reply .buildkite/pipeline.build.yml:669]: I've chatted with @kouroshHakha and those were the tests he initially suggested. Generally it's ok to add more tests here, but it would be good to keep the area relatively small. RLModule + examples will add another 9 test jobs.

Of course, if this is required we should do it, but generally the idea is that "Legacy support" means we ensure basic functionality, but not necessarily compatibility with latest features.

**Comment by krfricke** [comment]
> > However, this undermines our flaky test tracking, as actually flaky tests will be detected less often.
> 
> Are you saying go/flaky will be affected by this? like as long as one run succeeds, it's green on go/flaky?

Yes.

To be clear:
This is only for the tests with `pytest --reruns 3`. Generally this means that pytest will immediately re-run a subtest (e.g. `test_module.py::test_subtest` up to three times when it fails. Thus, if a test in those suites is _actually_ flaky, pytest will run it up to 3 times, if it passes, it will have passed for bazel and thus show up as green.

File: .buildkite/pipeline.build.yml
Comment (ArturNiederfahrenhorst) [reply .buildkite/pipeline.build.yml:669]: Got it, thanks 🙁

File: docker/retag-lambda/cuda_versions.txt
Comment (aslonnie) [inline docker/retag-lambda/cuda_versions.txt]: why are we back supporting cu115 and cu117? are there users that require those?

I guess the question is more like, why we started supporting cu116 and cu118 without supporting cu115 and cu117?

File: python/ray/air/tests/test_integration_mlflow.py
Comment (aslonnie) [inline python/ray/air/tests/test_integration_mlflow.py:46]: why is this change to `sqlite:///` required in this PR?

File: python/ray/tune/tests/execution/test_controller_control_integration.py
Comment (aslonnie) [inline python/ray/tune/tests/execution/test_controller_control_integration.py:137]: why do we need these `--reruns` ?

File: python/ray/serve/BUILD
Comment (zcin) [reply python/ray/serve/BUILD:598]: Ah I think our CI used python 3.6 when I wrote these tests and gradio required >= 3.7. I don't think there's anything blocking us from using 3.8 here

**Review by aslonnie — COMMENTED** [review]
overall looks good.

- I do not understand why some of the changes are required for upgrading python/cuda
- I also wonder if we really need to support cu115/cu117. like can we just support cu116/cu118? do we have evidence that users need cu115/cu117?

File: docker/retag-lambda/cuda_versions.txt
Comment (krfricke) [reply docker/retag-lambda/cuda_versions.txt]: This is due to the criteria we laid out in the support framework https://github.com/ray-project/ray/issues/34863

Generally we said that we'll support the CUDA versions that are also actively maintained by nvidia. This includes CUDA 11.5 and 11.7. I believe the main reason we didn't add them back then was because we already had a large number of GPU images and didn't want to inflate it further. I don't think it was a conscious decision - I think we had requests for 11.6 and later for 11.8, so we added them.

What we should have done is followed CUDA's lifecycle earlier and deprecated those images earlier. 

I'm also happy to leave those images out, but we'll have to find a good reasoning for this. I mainly care to have a plan going forward and a clear indication on which CUDA versions to add, keep, or remove, and following the CUDA image lifecycle seems easiest in terms of management.

File: python/ray/air/tests/test_integration_mlflow.py
Comment (krfricke) [reply python/ray/air/tests/test_integration_mlflow.py:46]: mlflow has been updated from 1.x to 2.x (2.x is not built for Py 3.7). There's been a breaking API change in that version switch, in that the local file backend now creates random experiment IDs instead of sequential ones starting from 0. The sqlite backend still starts from 0 (though this now points to a default experiment - thus the experiments we create in the tests start at ID 1). In order to keep changes to the tests minimal, switching the backend was an easy solution.

File: python/ray/tune/tests/execution/test_controller_control_integration.py
Comment (krfricke) [reply python/ray/tune/tests/execution/test_controller_control_integration.py:137]: This is described in the PR description:

> Due to a bug in libstdc++ (e.g. tracked here: https://github.com/ray-project/ray/issues/33858), the Ray processes sometimes error out. It looks like this occurred rarely in Python 3.7, but after the switch to Python 3.8 we see it more often in some of our tests.
>
> Note that this is an ephemeral failure and not due to specific tests. For instance, we saw test_trial_runner_3.py fail regularly, but it was always different tests. It seemed that just by the sheer amount of tests, and the quick succession of starting/stopping Ray clusters, one of these tests was bound to fail.
>
> For those tests, we introduced an intermediate solution using pytest reruns, namely the --reruns 3 option. This specifies reruns on the subtest level, as opposed to bazels test-suite level. The reruns option was able to make these flaky tests green. However, this undermines our flaky test tracking, as actually flaky tests will be detected less often. After moving to Ubuntu 22 and a new stdlibc++, we should remove these subtest-reruns.

File: python/ray/serve/BUILD
Comment (krfricke) [reply python/ray/serve/BUILD:598]: This is also my understanding

**Review by kouroshHakha — APPROVED** [review]

File: rllib/examples/bandit/tune_lin_ucb_train_recommendation.py
Comment (aslonnie) [inline rllib/examples/bandit/tune_lin_ucb_train_recommendation.py:98]: what change caused this change?

File: python/requirements/ml/requirements_upstream.txt
Comment (aslonnie) [inline python/requirements/ml/requirements_upstream.txt:11]: these are the same versions? why need the version split?

File: python/requirements/ml/requirements_tune.txt
Comment (aslonnie) [inline python/requirements/ml/requirements_tune.txt:7]: why is `sqlalchemy` required?

**Review by aslonnie — COMMENTED** [review]
I do not think I am the owner of any of these changes, and you probably do not need my approval. I am generally confused about why many of the changes are required for python version change. I feel that they at least need more explanations to justify.

for example, if we need to change some of the example code when upgrade from python 3.7 to 3.8. should we worry that the examples do not work on python 3.9, 3.10, which we claim to support and have container images that are released already?

I think most of the changes are not really due to python interpreter upgrade, but mostly due to library version changes of ray's dependencies. because the dependency version resolving for each python version will be different for each library (due to the `python_version` constraints in `requirements.txt` files), how do we know if ray works on python 3.9 / 3.10 ?

File: python/ray/data/tests/test_ecosystem.py
Comment (krfricke) [reply python/ray/data/tests/test_ecosystem.py:116]: It's because of https://github.com/ray-project/ray/pull/30895

- #30895 removes the worker init
- modin in the pinned version needs that API to work
- modin in the latest version does not need that API anymore, but will require a more recent pandas, which will break other tests

I can add a comment

File: python/ray/tests/test_cli.py
Comment (krfricke) [reply python/ray/tests/test_cli.py:283]: It looks like detection of interactive terminals works differently in Python 3.8. We can resolve this, but it should be done in a follow-up.

**Review by edoakes — APPROVED** [review]

File: python/ray/air/tests/test_integration_mlflow.py
Comment (krfricke) [reply python/ray/air/tests/test_integration_mlflow.py:46]: We shipped a later mlflow version in previous Ray releases for python 3.8+ and some of our customers/users rely on it. Pinning it to the lower version here would downgrade it in these images.

File: python/ray/tune/tests/execution/test_controller_control_integration.py
Comment (krfricke) [reply python/ray/tune/tests/execution/test_controller_control_integration.py:137]: We can get rid of the --reruns when we upgrade to ubuntu 22 (and newer stdlibc++). Instead of adding a flag that we then never set, let's just remove these alltogether in the follow-up where we migrate ubuntu.

Yes, the tests have always been "broken" on 3.8. Or to be clear, the tests are actually fine, it's just that the chance of running into the stdlibc++ race condition seems high in those particular tests. But yes, if we had ran them on 3.8 before, they would have popped up as red.

File: python/requirements/ml/requirements_tune.txt
Comment (krfricke) [reply python/requirements/ml/requirements_tune.txt:53]: This is the latest available version for python 3.7. It looks like later scipy versions are incompatible with the pinned dragonfly/bayesopt versions. this should be resolved in a follow-up.

File: rllib/examples/bandit/tune_lin_ucb_train_recommendation.py
Comment (krfricke) [reply rllib/examples/bandit/tune_lin_ucb_train_recommendation.py:98]: df.append has been deprecated and removed in e.g. pandas 1.5.x, which will be installed in some combinations of our requirements (1.3.5 is the last py3.7 release and still supported the deprecated API)

File: python/requirements/ml/requirements_tune.txt
Comment (krfricke) [reply python/requirements/ml/requirements_tune.txt:18]: IIRC newer versions (which are available for py 3.8) would lead some tests to fail

File: python/ray/air/tests/test_integration_mlflow.py
Comment (krfricke) [reply python/ray/air/tests/test_integration_mlflow.py:46]: I can separate this out though to make it clear. We will just have some nightly dockers with an older mlflow built then.

**Review by scv119 — APPROVED** [review]

File: python/ray/data/tests/test_ecosystem.py
Comment (bveeramani) [reply python/ray/data/tests/test_ecosystem.py:116]: Comment about this would be appreciated!

**Review by bveeramani — APPROVED** [review]
LGTM! If you could add a comment about the Modin dependency issue, that'd be awesome.

File: python/ray/air/tests/test_integration_mlflow.py
Comment (krfricke) [reply python/ray/air/tests/test_integration_mlflow.py:46]: I've separated the test change and will re-upgrade mlflow in a follow-up PR

**Review by richardliaw — APPROVED** [review]
overall looks good

**Review by amogkam — COMMENTED** [review]
Can we also update the corresponding documentation (installing ray guide) and the dockerhub info that the default python version is now 3.8 and not 3.7

**Review by alanwguo — APPROVED** [review]

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**
